### PR TITLE
Upgrade to latest mypy

### PR DESCRIPTION
**Pull request by achidlow**
PEP-484 has changed to recommend against implicit optionals based on default arguments, and mypy has accordingly flipped the default in a recent release.

ref: https://peps.python.org/pep-0484/#union-types

### CODE REVIEW

**Comment by barnjamin** [comment]
@tzaffi I recall seeing a recent commit from you switching the `thing | None` type hint to `Optional[thing]`, was there a specific reason for that?

**Comment by tzaffi** [comment]
> @tzaffi I recall seeing a recent commit from you switching the `thing | None` type hint to `Optional[thing]`, was there a specific reason for that?

These are [semantically equivalent](https://peps.python.org/pep-0604/). I think we ought to come to an agreement about what style we ought to adopt going forward -so we don't need to spend any mental calories on it. I propose:

* use `Optional[TheType]` instead of `TheType | None
* Use `TheFirstType | TheSecondType` instead of `Union[TheFirstType, TheSecondType]`

Why? I don't really have a good reason. I could be easily swayed to another viewpoint. But we should adopt _some_ style.

File: requirements.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@ black==22.3.0
 flake8==5.0.4
 flake8-tidy-imports==4.6.0
 graviton@git+https://github.com/algorand/graviton@v0.5.0
-mypy==0.950
+mypy==0.991
```
Comment (tzaffi) [inline requirements.txt:5]: 👍

**Comment by barnjamin** [comment]
@achidlow noted that the use of Union/Optional are required if the type must be quoted

**Comment by tzaffi** [comment]
> @achidlow noted that the use of Union/Optional are required if the type must be quoted

That's an argument for not using `|` at all.

File: mypy.ini [hunk 2]
```diff
@@ -1,4 +1,6 @@
 [mypy]
+files = pyteal,scripts,tests
```
Comment (tzaffi) [inline mypy.ini:2]: Should we remove the `MYPY` variable in the `Makefile` as well, now that the config is taking over this decision?

**Review by tzaffi — COMMENTED** [review]
Thanks for upgrading mypy!

I had a small nit regarding the `Makefile`.

Also, I won't hold up the PR on account of this, but I created #619 and suggested that we disallow `|` for annotations going forward.

File: mypy.ini (hunk 2 above)
Comment (achidlow) [reply mypy.ini:2]: Yes, that makes sense. Done ✅

**Comment by achidlow** [comment]
@tzaffi pending the outcome of #619, I've at least made this PR internally consistent in the use of `|` over `Optional`. If you'd prefer it the other way around for now I'm happy to flip that for all the methods changed in this PR.

File: Makefile [hunk 3]
```diff
@@ -40,9 +40,8 @@ black:
 flake8:
 	flake8 $(ALLPY)
 
-MYPY = pyteal scripts tests
 mypy:
-	mypy --show-error-codes $(MYPY)
+	mypy
```
Comment (tzaffi) [inline Makefile:44]: Why drop the `--show-error-codes` arg? It's been handy on occasion.

Comment (achidlow) [reply Makefile:44]: Not sure if this got flipped at some point, but that's [the default](https://mypy.readthedocs.io/en/stable/error_codes.html#displaying-error-codes) and assumed. You have to disable it with `--hide-error-codes`

**Review by tzaffi — APPROVED** [review]
Looks great, thank you!

**Pull request merged**